src/common/tensor_inspector.h

```diff
@@ -0,0 +1,762 @@
+      }
+      *os << tb_.dptr<DType>()[i];
+    }
+    *os << std::string(dimension, ']') << std::endl;
+    tensor_info_to_string<DType>(os);
+  }
+
+  /*!
+   * \brief output the tensor in a structured format 
+   * \tparam DType the data type
+   * \tparam StreamType the type of the stream object
+   * \param os stream object to output to
+   * \param dptr the data pointer
+   */
+  template<typename DType, typename StreamType>
+  void to_string_helper(StreamType* os, const DType* dptr) {
+#if MXNET_USE_CUDA
+    if (tb_.dev_mask() == gpu::kDevMask) {
+      TensorInspector(test::CAccessAsCPU(ctx_, tb_, false)(), ctx_)
+          .to_string_helper<DType>(os, dptr);
+      return;
+    }
+#endif  // MXNET_USE_CUDA
+    *os << *dptr << std::endl;
+    *os << "<" << typeid(*dptr).name() << ">" << std::endl;
+  }
+
+  /*!
+   * \brief output a part of the tensor in a structed format 
+   * \tparam DType the data type
+   * \tparam StreamType the type of the stream object
+   * \param os stream object to output to
+   * \param sub_shape the sub-shape of the desired part of the tensor
+   * \param offset the position of the first value of the desired part of the tensor
+   */
+  template<typename DType, typename StreamType>
+  void to_string_helper(StreamType* os, const std::vector<int>& sub_shape, index_t offset) {
+#if MXNET_USE_CUDA
+    if (tb_.dev_mask() == gpu::kDevMask) {
+      TensorInspector(test::CAccessAsCPU(ctx_, tb_, false)(), ctx_)
+          .to_string_helper<DType>(os, sub_shape, offset);
+      return;
+    }
+#endif  // MXNET_USE_CUDA
+    DType* dptr = tb_.dptr<DType>() + offset;
+    if (sub_shape.size() == 0) {
+      to_string_helper<DType>(os, dptr);
+      return;
+    }
+    const int dimension = sub_shape.size();
+    std::vector<index_t> multiples;
+    index_t multiple = 1;
+    for (int i = dimension-1; i >= 0; --i) {
+      multiple *= sub_shape[i];
+      multiples.push_back(multiple);
+    }
+    std::stringstream ss;
+    *os << std::string(dimension, '[');
+    *os << dptr[0];
+    for (index_t i = 1; i < multiple; ++i) {
+      int n = 0;
+      for (auto divisor : multiples) {
+        n += (i % divisor == 0);
+      }
+      if (n) {
+        *os << std::string(n, ']') << ", " <<  std::string(n, '[');
+      } else  {
+        *os << ", ";
+      }
+      *os << dptr[i];
+    }
+    *os << std::string(dimension, ']') << std::endl;
+    tensor_info_to_string<DType>(os, sub_shape);
+  }
+
+  /*!
+   * \brief helper function to calculate the sub_shape and offset for the desired part of the tensor,
+   * given its coordinates in the original tensor
+   * \param pos the coordinates of the desired part of the tensor
+   * \param sub_shape the sub-shape of the desired part of the tensor; calculated here
+   * \param offset the position of the first value of the desired part of the tensor; calculated here
+   */
+  void print_locator(const std::vector<int>& pos, std::vector<int>* sub_shape, index_t* offset) {
```

Comment:
Use index_t instead of int for pos vector. Essentially index to an element inside a tensor should have data type `index_t`

Reply:
good catch!